### PR TITLE
Fix for SNAP-3261.

### PR DESCRIPTION
**Pull request by smahajan05**
## Changes proposed in this pull request

Earlier code was wrong as it was removing the metrics for table name and was taking null value after the table gets deleted. So I modified code to remove table metrics only once and not generate the null value.
 
Added some test changes in dunit test
- Added code to match the table metrics values with the expected values
- Added code to match member metrics values with the expected values

## Patch testing
Manuall tested. Ran hydra test

## ReleaseNotes.txt changes

(Does this change require an entry in ReleaseNotes.txt? If yes, has it been added to it?)

## Other PRs 

(Does this change require changes in other projects- store, spark, spark-jobserver, aqp? Add the links of PR of the other subprojects that are related to this change)

### CODE REVIEW

File: cluster/src/main/scala/io/snappydata/metrics/SnappyMetricsSystem.scala [hunk 1]
```diff
@@ -234,8 +234,10 @@ object SnappyMetricsSystem {
   def removeTableMetricsIfTableDeleted(stringToStats: Map[String, SnappyRegionStats]): Unit = {
     for (t <- tables) {
       if (!stringToStats.exists(x => x._1 == t)) {
-        SnappyTableMetrics.removeTableMetrics(t)
-        tables -= t
+        if (t != null) {
```
Comment (ashetkar) [inline cluster/src/main/scala/io/snappydata/metrics/SnappyMetricsSystem.scala]: In which case, can the table-name be null? Is that a valid case or a bug?
If a valid case, it's better to check for null at the time of adding the entry itself in `tables`, in `setMetricsForTableStatDetails`?

File: cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala [hunk 2]
```diff
@@ -79,10 +71,10 @@ class SnappyMetricsSystemDUnitTest(s: String)
 
   override def afterClass(): Unit = {
     logInfo((snappyProductDir + "/sbin/snappy-stop-all.sh").!!)
-    s"rm -rf $snappyProductDir/work".!!
-    Files.deleteIfExists(Paths.get(snappyProductDir, "conf", "locators"))
-    Files.deleteIfExists(Paths.get(snappyProductDir, "conf", "leads"))
-    Files.deleteIfExists(Paths.get(snappyProductDir, "conf", "servers"))
+    // s"rm -rf $snappyProductDir/work".!!
```
Comment (ashetkar) [inline cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala]: What's the reason to retain the work directory?

File: cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala [hunk 3]
```diff
@@ -60,16 +53,15 @@ class SnappyMetricsSystemDUnitTest(s: String)
...
+        s"-dir=$snappyProductDir/work/lead2 -spark.ui.port=3334")
     pw1.close()
     val pw2 = new PrintWriter(new File(s"$confDir/servers"))
-    pw2.write(s"localhost -locators=localhost[$port] -client-port=$netPort2" +
-        s" -dir=$snappyProductDir/work/server1\n")
-    pw2.write(s"localhost -locators=localhost[$port] -client-port=$netPort3" +
-        s" -dir=$snappyProductDir/work/server2\n")
-    pw2.write(s"localhost -locators=localhost[$port] -client-port=$netPort4" +
-        s" -dir=$snappyProductDir/work/server3")
+    pw2.write(s"localhost -locators=localhost[$port] -dir=$snappyProductDir/work/server1\n")
```
Comment (ashetkar) [inline cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala]: Don't need to specify `-dir` property. Scripts automatically create these directories.

File: cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala [hunk 4]
```diff
@@ -60,16 +53,15 @@ class SnappyMetricsSystemDUnitTest(s: String)
         s" -peer-discovery-port=$port -client-port=$netPort")
     pw.close()
     val pw1 = new PrintWriter(new File(s"$confDir/leads"))
-    pw1.write(s"localhost -locators=localhost[$port] -dir=$snappyProductDir/work/lead1\n")
-    pw1.write(s"localhost -locators=localhost[$port] -dir=$snappyProductDir/work/lead2")
+    pw1.write(s"localhost -locators=localhost[$port] " +
```
Comment (ashetkar) [inline cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala]: Why specify `spark.ui.port` for each lead? Does it cause any conflict?

File: cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala [hunk 5]
```diff
@@ -119,13 +111,11 @@ class SnappyMetricsSystemDUnitTest(s: String)
...
-  def getConnection(): Connection = {
-    val driver = "io.snappydata.jdbc.ClientDriver"
-    Utils.classForName(driver).newInstance
-    DriverManager.getConnection(s"jdbc:snappydata:thrift://localhost[$netPort]")
-  }
+  def getConnection(netPort: Int): Connection =
+    DriverManager.getConnection(s"${Constant.DEFAULT_THIN_CLIENT_URL}localhost:$netPort")
 
   def testMetricsMonitoring(): Unit = {
+    Thread.sleep(10000)
```
Comment (ashetkar) [inline cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala]: Why is this sleep for and can we instead use the `WaitCriterion`?

File: cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala [hunk 6]
```diff
@@ -319,51 +309,35 @@ class SnappyMetricsSystemDUnitTest(s: String)
         assertEquals("APP.TEST2", v)}
       if (containsWords(k, Array("TableMetrics.APP.TEST2", "isReplicatedTable"))) {
         assertEquals(false, v)}
-      if (containsWords(k, Array("TableMetrics.APP.TEST2", "sizeSpillToDisk"))) {
```
Comment (ashetkar) [inline cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala]: Please explain these changes in the test as a PR description.

File: cluster/src/main/scala/io/snappydata/metrics/SnappyMetricsSystem.scala (hunk 1 above)
Comment (smahajan05) [reply cluster/src/main/scala/io/snappydata/metrics/SnappyMetricsSystem.scala]: Added the changes you suggested

File: cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala (hunk 2 above)
Comment (smahajan05) [reply cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala]: Mistakenly added comment for these lines. Removed it

File: cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala (hunk 4 above)
Comment (smahajan05) [reply cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala]: yes its causing conflict and i am getting bind exception So i added this spark.ui.port for getting metrics on that port

File: cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala (hunk 5 above)
Comment (smahajan05) [reply cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala]: Sleep is required to take effects in metrics values as the metrics values gets updated by 5 sec

File: cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala (hunk 6 above)
Comment (smahajan05) [reply cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala]: Explained the changes in pr description

File: cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala (hunk 3 above)
Comment (smahajan05) [reply cluster/src/dunit/scala/io/snappydata/metrics/SnappyMetricsSystemDUnitTest.scala]: I need that option for getting full and short directory names for members work directory.

**Pull request merged**